dos/src/main.rs

```diff
@@ -147,126 +306,173 @@ fn run_dos(
-    let data_type = value_t_or_exit!(matches, "data_type", String);
-    let data_input = value_t!(matches, "data_input", String).ok();
+/// input checks which are not covered by Clap
+fn validate_input(params: &DosClientParameters) {
+    if params.mode == Mode::Rpc
+        && (params.data_type != DataType::GetAccountInfo
+            && params.data_type != DataType::GetProgramAccounts)
+    {
+        panic!("unsupported data type");
+    }
+
+    if params.data_type != DataType::Transaction {
+        let tp = &params.transaction_params;
+        if tp.valid_blockhash
+            || tp.valid_signatures
+            || tp.unique_transactions
+            || tp.payer_filename.is_some()
+        {
+            println!("Arguments valid-blockhash, valid-sign, unique-trans, payer are ignored if data-type != transaction");
+        }
+    }
+
+    if params.transaction_params.payer_filename.is_some()
+        && params.transaction_params.valid_signatures
+    {
+        println!("Arguments valid-signatures is ignored if payer is provided");
+    }
+}
+
+fn main() {
+    solana_logger::setup_with_default("solana=info");
+    let cmd_params = DosClientParameters::parse();
+    validate_input(&cmd_params);
 
     let mut nodes = vec![];
-    if !skip_gossip {
-        info!("Finding cluster entry: {:?}", entrypoint_addr);
-        let socket_addr_space = SocketAddrSpace::new(matches.is_present("allow_private_addr"));
+    if !cmd_params.skip_gossip {
+        info!("Finding cluster entry: {:?}", cmd_params.entrypoint_addr);
+        let socket_addr_space = SocketAddrSpace::new(cmd_params.allow_private_addr);
         let (gossip_nodes, _validators) = discover(
             None, // keypair
-            Some(&entrypoint_addr),
-            None,                    // num_nodes
-            Duration::from_secs(60), // timeout
-            None,                    // find_node_by_pubkey
-            Some(&entrypoint_addr),  // find_node_by_gossip_addr
-            None,                    // my_gossip_addr
-            0,                       // my_shred_version
+            Some(&cmd_params.entrypoint_addr),
+            None,                              // num_nodes
+            Duration::from_secs(60),           // timeout
+            None,                              // find_node_by_pubkey
+            Some(&cmd_params.entrypoint_addr), // find_node_by_gossip_addr
+            None,                              // my_gossip_addr
+            0,                                 // my_shred_version
             socket_addr_space,
         )
         .unwrap_or_else(|err| {
-            eprintln!("Failed to discover {} node: {:?}", entrypoint_addr, err);
+            eprintln!(
+                "Failed to discover {} node: {:?}",
+                cmd_params.entrypoint_addr, err
+            );
             exit(1);
         });
         nodes = gossip_nodes;
     }
 
     info!("done found {} nodes", nodes.len());
+    let payer = cmd_params
```

Comment:
A comment about this complicated construction.
It is done this way to support 3 scenarios (I couldn't find a more elegant way):

1. when we run it as a tool we read payer keypair from file. Hence we, as a tool program, own the keypair
2. when we run in tests, we pass payer from cluster. So we don't own the keypair but borrow it by reference. keypair doesn't have clone
3. None





[9:44](https://solanalabs.slack.com/archives/D034KN9QTPF/p1647031489800329)
3. There is no payer, None